Lib/test/test_warnings/__init__.py

```diff
@@ -794,6 +794,33 @@ def test_stderr_none(self):
         self.assertNotIn(b'Warning!', stderr)
         self.assertNotIn(b'Error', stderr)
 
+    def test_warn_explicit(self):
+        # Issue #31285: warn_explicit() should neither raise a SystemError nor
+        # cause an assertion failure, in case the return value of get_source()
+        # is invalid.
+        def _get_bad_loader(splitlines_ret_val):
+            class BadLoader:
+                def get_source(self, fullname):
+                    class BadSource(str):
+                        def splitlines(self):
+                            return splitlines_ret_val
+                    return BadSource()
+            return BadLoader()
+        self.assertRaises(IndexError, self.module.warn_explicit,
```

Comment:
Since this test already takes multiple lines, I think it would look clearer if use a context manager form:
```
with self.assertRaises(IndexError):
     self.module.warn_explicit(...)
```

But why `IndexError` is expected? Wouldn't be better to test the normal case that doesn't raise an error? You can use `test.support.captured_stderr()` for capturing the warning output.